public/app/plugins/datasource/loki/live_streams.test.ts

```diff
@@ -0,0 +1,207 @@
+import { Subject, Observable } from 'rxjs';
+import * as rxJsWebSocket from 'rxjs/webSocket';
+import { LiveStreams } from './live_streams';
+import { DataFrameView, Labels, formatLabels, DataFrame } from '@grafana/data';
+import { noop } from 'lodash';
+
+let fakeSocket: Subject<any>;
+jest.mock('rxjs/webSocket', () => {
+  return {
+    __esModule: true,
+    webSocket: () => fakeSocket,
+  };
+});
+
+describe('Live Stream Tests', () => {
+  afterAll(() => {
+    jest.restoreAllMocks();
+  });
+
+  const msg0: any = {
+    streams: [
+      {
+        labels: '{filename="/var/log/sntpc.log", job="varlogs"}',
+        entries: [
+          {
+            ts: '2019-08-28T20:50:40.118944705Z',
+            line: 'Kittens',
+          },
+        ],
+      },
+    ],
+    dropped_entries: null,
+  };
+
+  it('reads the values into the buffer', done => {
+    fakeSocket = new Subject<any>();
+    const labels: Labels = { job: 'varlogs' };
+    const target = makeTarget('fake', labels);
+    const stream = new LiveStreams().getStream(target);
+    expect.assertions(5);
+
+    const tests = [
+      (val: DataFrame[]) => expect(val).toEqual([]),
+      (val: DataFrame[]) => {
+        expect(val[0].length).toEqual(7);
+        expect(val[0].labels).toEqual(labels);
+      },
+      (val: DataFrame[]) => {
+        expect(val[0].length).toEqual(8);
+        const view = new DataFrameView(val[0]);
+        const last = { ...view.get(view.length - 1) };
+        expect(last).toEqual({
+          ts: '2019-08-28T20:50:40.118944705Z',
+          line: 'Kittens',
+          labels: { filename: '/var/log/sntpc.log' },
+        });
+      },
+    ];
+    stream.subscribe({
```

Comment:
Ugh this took me longer than I would like. @ryantxu I changed your approach here where you used toPromise to just get the values from stream, because with the change to how the streams are made they are no hot anymore, meaning somebody has to subscribe to the stream. Though `onPromise` subscribes it subscribes from the start again. So I needed single subscription and at the same time I cannot just use `toPromise` on the whole stream because I get only references to single dataFrame and need to test intermediate states. That results in this ugly duckling of a test.

@hugohaggmark Any pointers here on some more sensible testing? I seen the marble tests but to be honest (without trying them) I do not like how they look but maybe that's the most sensible way to do this?

Reply:
I've tried to avoid marbe tests in the past, but in this newer version they might help. I usually subscribe to the whatever is the output and then expect that each data received in that subscribe is correct. Timing and scheduling is hard though.